cpp/inc/bond/protocol/detail/protobuf_utils.h

```diff
@@ -0,0 +1,299 @@
+                return type == BT_INT64 ? WireType::Fixed64 : WireType::Fixed32;
+
+            case Encoding::ZigZag:
+            default:
+                return WireType::VarInt;
+            }
+            break;
+
+        case BT_STRUCT:
+        case BT_STRING:
+            return WireType::LengthDelimited;
+
+        case BT_WSTRING:
+            NotSupportedException("Wide string");
+            break;
+
+        default:
+            break;
+        }
+
+        BOOST_ASSERT(false);
+        return Unavailable<WireType>();
+    }
+
+    template <typename Unused = void>
+    struct AttributeName
+    {
+        static const std::string encode;
+        static const std::string encode_key;
+        static const std::string encode_value;
+        static const std::string pack;
+    };
+
+    template <typename Unused>
+    const std::string AttributeName<Unused>::encode = "ProtoEncode";
+
+    template <typename Unused>
+    const std::string AttributeName<Unused>::encode_key = "ProtoEncodeKey";
+
+    template <typename Unused>
+    const std::string AttributeName<Unused>::encode_value = "ProtoEncodeValue";
+
+    template <typename Unused>
+    const std::string AttributeName<Unused>::pack = "ProtoPack";
+
+    inline Encoding ReadEncoding(const std::string& name, const Metadata& metadata)
+    {
+        auto it = metadata.attributes.find(name);
+        if (it == metadata.attributes.end())
+        {
+            return Unavailable<Encoding>();
+        }
+
+        if (std::strcmp(it->second.c_str(), "Fixed") == 0)
+        {
+            return Encoding::Fixed;
+        }
+
+        if (std::strcmp(it->second.c_str(), "ZigZag") == 0)
+        {
+            return Encoding::ZigZag;
+        }
+
+        InvalidEnumValueException(name.c_str(), "Encoding");
+    }
+
+    inline Encoding ReadEncoding(BondDataType type, const Metadata& metadata, const std::string& name)
+    {
+        switch (type)
+        {
+        case BT_FLOAT:
+        case BT_DOUBLE:
+            return Encoding::Fixed;
+
+        case BT_UINT8:
+        case BT_UINT16:
+        case BT_UINT32:
+        case BT_UINT64:
+        case BT_INT8:
+        case BT_INT16:
+        case BT_INT32:
+        case BT_INT64:
+            return ReadEncoding(name, metadata);
+
+        case BT_WSTRING:
+        case BT_SET:
+        case BT_MAP:
+            NotSupportedException("Wide string or nested set/map");
+
+        case BT_BOOL:
+        case BT_LIST:
+        case BT_STRING:
+        case BT_STRUCT:
+        default:
+            return Unavailable<Encoding>();
+        }
+    }
+
+    inline Encoding ReadEncoding(BondDataType type, const Metadata& metadata)
+    {
+        return ReadEncoding(type, metadata, AttributeName<>::encode);
+    }
+
+    inline Encoding ReadKeyEncoding(BondDataType type, const Metadata& metadata)
+    {
+        if (type == BT_FLOAT || type == BT_DOUBLE)
+        {
+            NotSupportedException("Float/double map key");
```

Comment:
This is oddly specific. Why is this a problem?

Reply:
Restriction in protobuf: https://developers.google.com/protocol-buffers/docs/proto3#maps
> where the key_type can be any integral or string type (so, any scalar type except for floating point types and bytes)